### PR TITLE
Feat/bootstrap ci

### LINKED ISSUE #87: Confidence Interval Protocol + BootstrapConfidenceInterval
## Summary

Introduce a `ConfidenceInterval` structural protocol so that `SemiSupervisedMeanInferenceResult` can hold either a CLT-based or a bootstrap-based confidence interval interchangeably. Implement `BootstrapConfidenceInterval` as the second concrete type.

---

## Motivation

The existing `SemiSupervisedMeanInferenceResult` hard-codes `confidence_interval: CLTConfidenceInterval`. The upcoming `PTDMeanEstimator` produces bootstrap percentile intervals, which cannot be stored in a `CLTConfidenceInterval` (no `std`, non-Gaussian shape). Rather than duplicating the result type, we introduce a lightweight `ConfidenceInterval` protocol that both CI classes satisfy, and update the result type's annotation accordingly.

---

### 0. New package: `glide/confidence_intervals/`

All CI-related code (protocol + concrete classes) lives in a dedicated sub-package, mirroring the existing `glide/core/mean_inference_result/` layout:

```
glide/confidence_intervals/
├── __init__.py                    # re-exports: ConfidenceInterval, CLTConfidenceInterval, BootstrapConfidenceInterval
├── base.py                        # ConfidenceInterval protocol
├── clt.py                         # CLTConfidenceInterval (moved from glide/core/clt_confidence_interval.py)
└── bootstrap.py                   # BootstrapConfidenceInterval (new)
```

`glide/core/clt_confidence_interval.py` should be **removed** once all internal imports are updated to point to `glide.confidence_intervals`.

---

### 1. `glide/confidence_intervals/base.py`

Define a structural protocol that every CI class must satisfy:

```python
from typing import Protocol, Tuple

class ConfidenceInterval(Protocol):
    confidence_level: float
    mean: float
    std: float
    var: float

    @property
    def lower_bound(self) -> float: ...

    @property
    def upper_bound(self) -> float: ...

    def test_null_hypothesis(
        self,
        h0_value: float,
        alternative: str = "two-sided",
    ) -> Tuple[float, float, float]:
        """Return (test_statistic, p_value, df)."""
        ...
```

`mean` and `std` are required because `MeanInferenceResult.mean` and `MeanInferenceResult.std`
delegate to `confidence_interval.mean` and `confidence_interval.std` respectively
(see `glide/core/mean_inference_result/base.py`).

`CLTConfidenceInterval` satisfies all members except `var`, which must be added as a `field(init=False)` computed in `__post_init__` (see §2).

---

### 2. `glide/confidence_intervals/clt.py`

Move `CLTConfidenceInterval` from `glide/core/clt_confidence_interval.py` to this file. Update all internal imports accordingly.

As part of this move:
- Add a complete class docstring matching the quality of `BootstrapConfidenceInterval`: a one-line summary, a description of the CLT-based construction, `Parameters`, `Returns` and an `Examples` section with a minimal runnable snippet.
- Add a `var` field computed once in `__post_init__`:

```python
from dataclasses import dataclass, field

@dataclass
class CLTConfidenceInterval:
    mean: float
    std: float
    confidence_level: float = 0.95
    var: float = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.var = self.std ** 2
        self.z_score = norm.ppf((1 + self.confidence_level) / 2)  # Remove the `z_score` method
```

**Design note — primary quantity:** `CLTConfidenceInterval` stores `std` as its primary constructor field; `var` is computed once from it in `__post_init__`. `BootstrapConfidenceInterval` takes the opposite direction: `var = np.var(bootstrap_estimates, ddof=1)` is primary, and `std = sqrt(self.var)` is derived. In both cases `mean`, `var`, and `std` are plain float attributes after construction — no recomputation on access.

---

### 3. `glide/confidence_intervals/bootstrap.py`

```python
from dataclasses import dataclass, field
from typing import Tuple

import numpy as np
from numpy.typing import NDArray


@dataclass
class BootstrapConfidenceInterval:
    """Quantile bootstrap confidence interval.

    Stores the full distribution of bootstrap point estimates and derives
    bounds as quantiles of that distribution.

    Parameters
    ----------
    bootstrap_estimates : NDArray
        Array of shape (B,) containing the B bootstrap point estimates.
    confidence_level : float
        Target coverage, e.g. 0.95 for a 95 % CI.

    Examples
    --------
    >>> import numpy as np
    >>> from glide.confidence_intervals import BootstrapConfidenceInterval
    >>> rng = np.random.default_rng(0)
    >>> estimates = rng.normal(loc=5.0, scale=0.3, size=20)
    >>> ci = BootstrapConfidenceInterval(bootstrap_estimates=estimates, confidence_level=0.95)
    >>> print(ci.lower_bound, ci.upper_bound)
    4.412 5.588
    """

    bootstrap_estimates: NDArray
    confidence_level: float = 0.95
    mean: float = field(init=False, repr=False)
    var: float = field(init=False, repr=False)
    std: float = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.mean = float(np.mean(self.bootstrap_estimates))
        self.var = float(np.var(self.bootstrap_estimates, ddof=1))
        self.std = float(np.sqrt(self.var))

    @property
    def lower_bound(self) -> float:
        tail = (1 - self.confidence_level) / 2
        bound = float(np.quantile(self.bootstrap_estimates, tail))
        return bound

    @property
    def upper_bound(self) -> float:
        tail = (1 - self.confidence_level) / 2
        bound = float(np.quantile(self.bootstrap_estimates, 1 - tail))
        return bound

    def test_null_hypothesis(
        self,
        h0_value: float,
        alternative: str = "two-sided",
    ) -> Tuple[float, float, float]:
        """Bootstrap hypothesis test against a null value.

        Computes a p-value as the proportion of bootstrap estimates that are
        at least as extreme as `h0_value` under the specified alternative.

        Parameters
        ----------
        h0_value : float
            Hypothesised population mean under H0.
        alternative : str, optional
            One of ``'two-sided'``, ``'larger'``, or ``'smaller'``.

        Returns
        -------
        Tuple[float, float, float]
            ``(test_statistic, p_value, df)`` where ``test_statistic`` is the
            point estimate (mean of bootstrap distribution), ``p_value`` is the
            bootstrap p-value, and ``df`` is ``float('inf')``.
        """
        if alternative == "two-sided":
            centered = np.abs(self.bootstrap_estimates - self.mean)
            observed_deviation = abs(self.mean - h0_value)
            is_at_least_as_extreme = centered >= observed_deviation
            p_value = float(np.mean(is_at_least_as_extreme))
        elif alternative == "larger":
            is_at_least_as_extreme = self.bootstrap_estimates <= h0_value
            p_value = float(np.mean(is_at_least_as_extreme))
        elif alternative == "smaller":
            is_at_least_as_extreme = self.bootstrap_estimates >= h0_value
            p_value = float(np.mean(is_at_least_as_extreme))
        else:
            raise ValueError(
                f"alternative must be 'two-sided', 'larger', or 'smaller', got '{alternative}'"
            )
        return self.mean, p_value, float("inf")
```

---

### 4. Update `glide/core/mean_inference_result/semi_supervised.py`

Change the type annotation of `confidence_interval` from `CLTConfidenceInterval` to `ConfidenceInterval`:

```python
# Before
from glide.core.clt_confidence_interval import CLTConfidenceInterval

@dataclass(repr=False)
class SemiSupervisedMeanInferenceResult(MeanInferenceResult):
    confidence_interval: CLTConfidenceInterval

# After
from glide.confidence_intervals import ConfidenceInterval

@dataclass(repr=False)
class SemiSupervisedMeanInferenceResult(MeanInferenceResult):
    confidence_interval: ConfidenceInterval
```

The `_common_lines()` method in `MeanInferenceResult` (base class) accesses `confidence_interval.lower_bound`, `confidence_interval.upper_bound`, and `confidence_interval.confidence_level` — all of which are part of the protocol, so no further changes are needed.

---

### 5. Simplify `glide/core/utils.py` — `compute_effective_sample_size`

The current implementation computes:

```python
std_of_mean = np.sqrt(var_y_true / n)
effective_sample_size = int(n * (std_of_mean / effective_std) ** 2)
```

Expanding the square shows that `n` cancels entirely:

```
n * (√(var_y_true / n) / effective_std)² = n * var_y_true / n / effective_std² = var_y_true / effective_std²
```

Adding a `var` property to every CI type (see §3) eliminates the squaring entirely. The simplified implementation takes `effective_var` rather than `effective_std`:

```python
def compute_effective_sample_size(y_true: np.ndarray, effective_var: float) -> int:
    var_y_true = np.nanvar(y_true, ddof=1)
    effective_sample_size = int(var_y_true / effective_var)
    return effective_sample_size
```

Call sites (PPI, ASI, PTD) pass `ci.var`:

```python
effective_sample_size = compute_effective_sample_size(y_true, ci.var)
```

`y_true` is still passed (rather than its variance) to keep the call site readable and consistent across estimators.

**Why `var` rather than `std`:** for `BootstrapConfidenceInterval`, `np.var` is the direct computation; deriving `std` via `sqrt` only to square it back at the call site is wasteful. For `CLTConfidenceInterval`, `var = std**2` is trivial. The formula `ESS = var(y_true) / effective_var` is also clearer in intent.

---

### 6. Update `glide/__init__.py` (if applicable)

Export the new classes from the `glide` namespace if other exports already exist there.

---

### 6. Update the architectural overview in the CONTRIBUTING.md

Enrich the current overview with the mean_inference_result folder

---

## Acceptance Criteria

- `BootstrapConfidenceInterval` is importable from `glide.confidence_intervals`.
- `SemiSupervisedMeanInferenceResult` can be instantiated with either a `CLTConfidenceInterval` or a `BootstrapConfidenceInterval` without type errors.
- `CLTConfidenceInterval` has a complete docstring (summary, `Parameters`, `Examples`) of equivalent quality to `BootstrapConfidenceInterval`.
- All existing tests for `PPIMeanEstimator` and `ASIMeanEstimator` continue to pass without modification.
- `compute_effective_sample_size(y_true, ci.std)` works identically for both CI types, enabling `PTDMeanEstimator` to compute effective sample size without any branching logic.
- `compute_effective_sample_size` is simplified to `int(np.nanvar(y_true, ddof=1) / effective_std**2)` — the intermediate `n` and `std_of_mean` variables are removed.

### PR DESCRIPTION
**Pull request by mraki-ed**
### Description

- What does this PR do?
Add ConfidenceInterval protocol and implements BootStrapConfidenceInterval
- Which issue does it close? (use `Closes #<number>`)
Closes #87 
- Any noteworthy implementation decisions or trade-offs?

### Type of change

Checkbox list:
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Refactoring (no behavior change)
- [ ] Documentation
- [ ] Repository hygiene

### Checklist

Quality gates that must be satisfied before requesting a review:

- [x] I have read `CONTRIBUTING.md`
- [x] `make lint` passes
- [x] `make type-check` passes
- [x] `make tests` passes
- [x] `make coverage` reports 100% coverage
- [x] New public API has numpy-style docstrings
- [x] New public API is inserted in the API reference section of the documentation
- [x] Docs build without warnings (`make doc`)
- [x] `CHANGELOG.md` updated if the change is user-facing

### LLM usage

Disclose if an LLM was used in writing this PR:
- [ ] No LLM used
- [x] LLM used: Claude
- [x] I went through and validated all the code myself

### CODE REVIEW

**Comment by mraki-ed** [comment]
### Code review

Found 3 issues:

1. One-sided test directions are swapped for `alternative='larger'` and `alternative='smaller'`. For `larger`, the code uses `<=` when it should use `>=`, and vice versa for `smaller`. This inverts the p-value direction for one-sided hypothesis tests, producing incorrect results.

https://github.com/EmertonData/glide/blob/50b64fcfc9d6fdb34e56a87ac8ce538be6d757e5/glide/confidence_intervals/bootstrap.py#L84-L87

2. Two-sided bootstrap p-value is conceptually incorrect. It centers the bootstrap distribution around `self.mean` instead of `h0_value`. The standard percentile bootstrap hypothesis test should shift the bootstrap distribution to be centered on `h0_value` and count the fraction that exceed the observed statistic, not center on the estimate itself.

https://github.com/EmertonData/glide/blob/50b64fcfc9d6fdb34e56a87ac8ce538be6d757e5/glide/confidence_intervals/bootstrap.py#L81-L83

3. Integer truncation behavior change in `compute_effective_sample_size` refactor. The old formula used `int()` on an intermediate calculation, while the new one changes the order of operations, silently changing results by 1 in edge cases (old test expected 24, new expects 25, for numerically equivalent inputs).

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: glide/confidence_intervals/base.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    @property
+    def upper_bound(self) -> float:
+        """Upper bound of the confidence interval."""
+        ...
+
+    def test_null_hypothesis(
+        self,
+        h0_value: float,
+        alternative: str = "two-sided",
```
Comment (gmartinon-ed) [inline glide/confidence_intervals/base.py]: Est-ce qu'on mettrait pas un `Literal` ici, avec les trois choix possibles ?

File: glide/confidence_intervals/base.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    """Structural protocol for confidence intervals.
+
+    Any class implementing this protocol can be used as a confidence_interval
+    in result objects like SemiSupervisedMeanInferenceResult.
+    """
+
+    confidence_level: float
+    mean: float
+    std: float
+    var: float
```
Comment (gmartinon-ed) [inline glide/confidence_intervals/base.py]: Je savais pas qu'on pouvait mettre des attributs dans un protocole. A quoi ça sert ? Est-ce qu'on est sûr que ces attributs ne sont pas obligatoires au moment de la construction ? Car tous les ConfidenceInterval ne sont pas initialisés avec ces 4 attributs.

File: tests/unit/core/test_utils.py [hunk 3]
```diff
@@ -5,6 +5,6 @@
 
 def test_compute_effective_sample_size_manual():
     y_true = np.array([5.0, 6.0, 7.0])
-    std = 0.2
-    effective_sample_size = compute_effective_sample_size(y_true, std)
-    assert effective_sample_size == 24
+    effective_var = 0.04  # equivalent to std=0.2
```
Comment (gmartinon-ed) [inline tests/unit/core/test_utils.py]: remove comment

File: tests/unit/io/test_to_json.py [hunk 4]
```diff
@@ -34,6 +34,7 @@ def test_to_json_semisupervised_mean_inference_result():
             "mean": pytest.approx(0.70, abs=1e-2),
             "std": pytest.approx(0.05, abs=1e-2),
             "confidence_level": pytest.approx(0.95, abs=1e-2),
+            "var": pytest.approx(0.0025, abs=1e-4),
```
Comment (gmartinon-ed) [inline tests/unit/io/test_to_json.py:37]: Pas sûr qu'on veuille afficher et la var et la std, seule std suffit je pense

File: glide/confidence_intervals/bootstrap.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            Hypothesised population mean under H0.
+        alternative : str, optional
+            One of ``'two-sided'``, ``'larger'``, or ``'smaller'``.
+
+        Returns
+        -------
+        Tuple[float, float, float]
+            ``(test_statistic, p_value, df)`` where ``test_statistic`` is the
+            point estimate (mean of bootstrap distribution), ``p_value`` is the
+            bootstrap p-value, and ``df`` is ``float('inf')``.
```
Comment (gmartinon-ed) [inline glide/confidence_intervals/bootstrap.py:81]: Reprendre les formulations de CLTConfidenceInterval

File: glide/confidence_intervals/bootstrap.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        if alternative == "two-sided":
+            centered = np.abs(self.bootstrap_estimates - self.mean)
+            observed_deviation = abs(self.mean - h0_value)
```
Comment (gmartinon-ed) [inline glide/confidence_intervals/bootstrap.py]: Amour de la symétrie
```suggestion
            observed_deviation = abs(h0_value - self.mean)
```

File: tests/unit/confidence_intervals/test_bootstrap.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    assert ci.lower_bound == pytest.approx(expected_lower, abs=0.001)
+
+
+def test_upper_bound(estimates):
+    ci = BootstrapConfidenceInterval(bootstrap_estimates=estimates, confidence_level=0.95)
+    expected_upper = 4.388
+    assert ci.upper_bound == pytest.approx(expected_upper, abs=0.001)
+
+
+# --- test_null_hypothesis ---
```
Comment (gmartinon-ed) [inline tests/unit/confidence_intervals/test_bootstrap.py:44]: J'aime ces tests concis et peu nombreux. Possible d'aligner les tests de CLT ? Y compris le naming ?

File: tests/unit/confidence_intervals/test_bootstrap.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def test_null_hypothesis_null_is_true(estimates):
+    ci = BootstrapConfidenceInterval(bootstrap_estimates=estimates)
+    _, p_value, df = ci.test_null_hypothesis(h0_value=2.25)
+    assert p_value == pytest.approx(1.0, abs=0.001)
+    assert df == float("inf")
+
+
+def test_null_hypothesis_two_sided(estimates):
```
Comment (gmartinon-ed) [inline tests/unit/confidence_intervals/test_bootstrap.py:54]: J'aime ce naming simple.
Possible d'aligner le naming de CLT en regard ?

**Review by gmartinon-ed — CHANGES_REQUESTED** [review]

File: glide/confidence_intervals/base.py (hunk 1 above)
Comment (imerad) [reply glide/confidence_intervals/base.py]: Traité

File: glide/confidence_intervals/base.py (hunk 2 above)
Comment (imerad) [reply glide/confidence_intervals/base.py]: Malgré qu'on a défini un protocole. Il n'y a pas de mécanisme de vérification que ce dernier soit respecté et ça ne se fait effectivement pas. Le fait de mentionner ces attributs ici indique quand même qu'ils sont censés exister dans un ConfidenceInterval. C'était comme ça dans le ticket aussi.

File: tests/unit/core/test_utils.py (hunk 3 above)
Comment (imerad) [reply tests/unit/core/test_utils.py]: Fait

File: tests/unit/io/test_to_json.py (hunk 4 above)
Comment (imerad) [reply tests/unit/io/test_to_json.py:37]: Ce champ s'est rajouté automatiquement avec le nouvel attribut var dans les ConfidenceInterval. Il faudrait surcharger la méthode asdict pour l'éviter ... Est ce qu'on veut vraiment le faire ?

File: glide/confidence_intervals/bootstrap.py (hunk 6 above)
Comment (imerad) [reply glide/confidence_intervals/bootstrap.py]: fait

File: glide/confidence_intervals/bootstrap.py (hunk 5 above)
Comment (imerad) [reply glide/confidence_intervals/bootstrap.py:81]: fait

File: tests/unit/confidence_intervals/test_bootstrap.py (hunk 7 above)
Comment (imerad) [reply tests/unit/confidence_intervals/test_bootstrap.py:44]: Les tests de CLT sont sur mesure pour la loi gaussienne et font référence au z_score ce qui ne peut pas être adapté pour les tests de bootstrap. Donc c'est à priori pas évident de les réconcilier

File: tests/unit/confidence_intervals/test_bootstrap.py (hunk 8 above)
Comment (imerad) [reply tests/unit/confidence_intervals/test_bootstrap.py:54]: fait

File: glide/confidence_intervals/base.py (hunk 2 above)
Comment (gmartinon-ed) [reply glide/confidence_intervals/base.py]: OK, dans ce cas je pense qu'on peut les supprimer

File: tests/unit/io/test_to_json.py (hunk 4 above)
Comment (gmartinon-ed) [reply tests/unit/io/test_to_json.py:37]: Ok pas nécessaire

File: glide/confidence_intervals/bootstrap.py (hunk 5 above)
Comment (gmartinon-ed) [reply glide/confidence_intervals/bootstrap.py:81]: Je ne crois pas ?

File: glide/confidence_intervals/base.py (hunk 2 above)
Comment (imerad) [reply glide/confidence_intervals/base.py]: On a seulement pu supprimer `var`. Les autres doivent rester car utilisés dans `glide/core/mean_inference_result/base.py`

File: glide/confidence_intervals/bootstrap.py (hunk 5 above)
Comment (imerad) [reply glide/confidence_intervals/bootstrap.py:81]: J'ai initialement cru que tu voulais dire de structurer cette partie de la même manière, mais on ne peut pas dire la même chose ici car le test d'hypothèse dans bootstrap fonctionne différemment. Le test de CLT évalue la probabilité d'observer la déviation constatée pour une variable Gaussienne standard. Le test de bootstrap lui fait la même chose mais pour une variable uniforme sur $[0, 1]$

Comment (gmartinon-ed) [reply glide/confidence_intervals/bootstrap.py:81]: On peut pas écrire ?

```python
"""
Parameters
        ----------
        h0_value : float
            The hypothesized population mean under the null hypothesis (H0: μ = h0_value).
        alternative : str, optional
            The alternative hypothesis. One of:
            - ``'two-sided'`` (default): H1: μ ≠ h0_value
            - ``'larger'``: H1: μ > h0_value
            - ``'smaller'``: H1: μ < h0_value
 """
```

Comment (imerad) [reply glide/confidence_intervals/bootstrap.py:81]: Ah, j'avais mal compris ... C'est fait

**Review by gmartinon-ed — APPROVED** [review]
